pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/OverloadShedder.java

```diff
@@ -50,63 +55,82 @@
-                double maxMessageRate = Double.NEGATIVE_INFINITY;
-                String mostTaxingBundle = null;
-                if (localData.getBundles().size() > 1) {
-                    for (final String bundle : localData.getBundles()) {
-                        final BundleData bundleData = loadData.getBundleData().get(bundle);
-                        if (bundleData == null) {
-                            continue;
-                        }
-
-                        // Consider short-term message rate to address system resource burden
-                        final TimeAverageMessageData shortTermData = bundleData.getShortTermData();
-                        final double messageRate = shortTermData.getMsgRateIn() + shortTermData.getMsgRateOut();
-                        // The burden of checking the timestamp is for the load manager, not the strategy.
-                        if (messageRate > maxMessageRate && !recentlyUnloadedBundles.containsKey(bundle)) {
-                            maxMessageRate = messageRate;
-                            mostTaxingBundle = bundle;
-                        }
-                    }
-                    if (mostTaxingBundle != null) {
-                        selectedBundlesCache.put(broker, mostTaxingBundle);
-                    } else {
-                        log.warn("Load shedding could not be performed on broker {} because all bundles assigned to it "
-                                + "have recently been unloaded");
-                    }
-                } else if (localData.getBundles().size() == 1) {
-                    log.warn(
-                            "HIGH USAGE WARNING : Sole namespace bundle {} is overloading broker {}. "
-                                    + "No Load Shedding will be done on this broker",
-                            localData.getBundles().iterator().next(), broker);
-                } else {
-                    log.warn("Broker {} is overloaded despite having no bundles", broker);
+            final double currentUsage = localData.getMaxResourceUsage();
+            if (currentUsage < overloadThreshold) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Broker is not overloaded, ignoring at this point", broker);
                 }
+                return;
             }
-        }
+
+            // We want to offload enough traffic such that this broker will go below the overload threshold
+            // Also, add a small margin so that this broker won't be very close to the threshold edge.
+            double percentOfTrafficToOffload = currentUsage - overloadThreshold + ADDITIONAL_THRESHOLD_PERCENT_MARGIN;
+            double brokerCurrentThroughput = localData.getMsgThroughputIn() + localData.getMsgThroughputOut();
+
+            double minimumThroughputToOffload = brokerCurrentThroughput * percentOfTrafficToOffload;
```

Comment:
can we add unit-test for this method?

Reply:
Yes, that was my initial intention :) adding now